pkg/storage/kubernetes/ipam.go

```diff
@@ -200,54 +200,63 @@ func (i *KubernetesIPAM) GetOverlappingRangeStore() (storage.OverlappingRangeSto
+	networkName string) (*whereaboutsv1alpha1.OverlappingRangeIPReservation, error) {
 	normalizedIP := normalizeIP(ip, networkName)
 
 	logging.Debugf("OverlappingRangewide allocation check; normalized IP: %q, IP: %q, networkName: %q",
 		normalizedIP, ip, networkName)
 
-	_, err := c.client.WhereaboutsV1alpha1().OverlappingRangeIPReservations(c.namespace).Get(ctx, normalizedIP, metav1.GetOptions{})
+	reservation, err := c.client.WhereaboutsV1alpha1().OverlappingRangeIPReservations(c.namespace).Get(ctx, normalizedIP, metav1.GetOptions{})
 	if err != nil && errors.IsNotFound(err) {
 		// cluster ip reservation does not exist, this appears to be good news.
 		// logging.Debugf("IP %v is not reserved cluster wide, allowing.", ip)
-		return false, nil
+		return nil, nil
 	} else if err != nil {
-		logging.Errorf("k8s get OverlappingRangeIPReservation error: %s", err)
-		return false, fmt.Errorf("k8s get OverlappingRangeIPReservation error: %s", err)
+		return nil, logging.Errorf("k8s get OverlappingRangeIPReservation error: %v", err)
 	}
 
 	logging.Debugf("Normalized IP is reserved; normalized IP: %q, IP: %q, networkName: %q",
 		normalizedIP, ip, networkName)
-	return true, nil
+	return reservation, nil
 }
 
 // UpdateOverlappingRangeAllocation updates clusterwide allocation for overlapping ranges.
 func (c *KubernetesOverlappingRangeStore) UpdateOverlappingRangeAllocation(ctx context.Context, mode int, ip net.IP,
 	containerID, podRef, networkName string) error {
 	normalizedIP := normalizeIP(ip, networkName)
-
-	clusteripres := &whereaboutsv1alpha1.OverlappingRangeIPReservation{
-		ObjectMeta: metav1.ObjectMeta{Name: normalizedIP, Namespace: c.namespace},
+	clusteripres, err := c.client.WhereaboutsV1alpha1().OverlappingRangeIPReservations(c.namespace).Get(ctx, normalizedIP, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return logging.Errorf("k8s get OverlappingRangeIPReservation error: %v", err)
 	}
 
-	var err error
 	var verb string
 	switch mode {
 	case whereaboutstypes.Allocate:
 		// Put together our cluster ip reservation
 		verb = "allocate"
 
-		clusteripres.Spec = whereaboutsv1alpha1.OverlappingRangeIPReservationSpec{
-			ContainerID: containerID,
-			PodRef:      podRef,
+		if errors.IsNotFound(err) {
+			clusteripres = &whereaboutsv1alpha1.OverlappingRangeIPReservation{
+				ObjectMeta: metav1.ObjectMeta{Name: normalizedIP, Namespace: c.namespace},
+			}
+			clusteripres.Spec = whereaboutsv1alpha1.OverlappingRangeIPReservationSpec{
+				ContainerID: containerID,
+				PodRef:      podRef,
+			}
+			_, err = c.client.WhereaboutsV1alpha1().OverlappingRangeIPReservations(c.namespace).Create(
+				ctx, clusteripres, metav1.CreateOptions{})
+		} else {
+			clusteripres.Spec = whereaboutsv1alpha1.OverlappingRangeIPReservationSpec{
+				ContainerID: containerID,
+				PodRef:      podRef,
+			}
+			_, err = c.client.WhereaboutsV1alpha1().OverlappingRangeIPReservations(c.namespace).Update(
```

Comment:
wrap in retry?

Reply:
It will return an error and fail the CNI call. So kubelet will retry the CNI call. Additionally, we don't have the retry for other API calls either. Therefore, if we want to implement a retry logic, we want to address it comprehensively.

Reply:
yeah o.k. I misunderstood this anyway given that it's a CNI plugin it doesn't make a lot of sense